src/trio/_tests/type_tests/raisesgroup.py

```diff
@@ -46,29 +19,34 @@ def check_matcher_typevar_default(e: Matcher) -> object:
 
 
 def check_basic_contextmanager() -> None:
-    # One level of Group is correctly translated - except it's a BaseExceptionGroup
-    # instead of an ExceptionGroup.
+
     with RaisesGroup(ValueError) as e:
         raise ExceptionGroup("foo", (ValueError(),))
-    assert_type(e.value, BaseExceptionGroup[ValueError])
+    assert_type(e.value, ExceptionGroup[ValueError])
 
 
 def check_basic_matches() -> None:
     # check that matches gets rid of the naked ValueError in the union
     exc: ExceptionGroup[ValueError] | ValueError = ExceptionGroup("", (ValueError(),))
     if RaisesGroup(ValueError).matches(exc):
-        assert_type(exc, BaseExceptionGroup[ValueError])
+        assert_type(exc, ExceptionGroup[ValueError])
+
+    # also check that BaseExceptionGroup shows up for BaseExceptions
+    if RaisesGroup(KeyboardInterrupt).matches(exc):
+        assert_type(exc, BaseExceptionGroup[KeyboardInterrupt])
 
 
 def check_matches_with_different_exception_type() -> None:
-    # This should probably raise some type error somewhere, since
-    # ValueError != KeyboardInterrupt
     e: BaseExceptionGroup[KeyboardInterrupt] = BaseExceptionGroup(
         "",
         (KeyboardInterrupt(),),
     )
+
+    # note: it might be tempting to have this warn.
+    # however, note that we should not warn if e: BaseException
+    # .... therefore this needs to pass as there is no distinction
     if RaisesGroup(ValueError).matches(e):
```

Comment:
Because if we type `.matches` like `.matches(e: <TypeVar bound to ValueError>)` then `.matches(BaseException())` will fail.

Reply:
hmm, I think this might need higher-kinded typevars or https://github.com/python/typing/issues/1226 in order to express the relationship between `self` and `exc_val`. So yeah probably just update the comment and say it's a limitation in the typing system.

Reply:
I'm not sure even then because it would be weird if this was allowed:
```
exc: BaseException = e
RaisesGroup(OtherException).matches(exc)
```

But this failed:
```
RaisesGroup(OtherException).matches(e)
```

... it's much more on the lint rather than type check side of the scale.

Reply:
I'm not sure, when broadening you're giving the type checker more leeway to match things and prevent it from catching errors. Code like `if isinstance(e, ValueError)` will be considered `unreachable` if e is narrow but will work with `e: BaseException`. I agree it's on the linting end of typing, but why wouldn't you want to help lint with type hints.

But that's mostly academical, a comment like "This cannot be made to warn with current typing, and not entirely sure it would be correct to do so" should cover all bases.